nlmod/dims/layers.py

```diff
@@ -807,77 +963,134 @@ def fill_nan_top_botm_kh_kv(
     Steps:
 
     1. Compute top and botm values, by filling nans by data from other layers
-    2. Compute idomain from the layer thickness
+    2. Remove inactive layers, with no positive thickness anywhere
     3. Compute kh and kv, filling nans with anisotropy or fill_values
     """
 
     # 1
     ds = fill_top_and_bottom(ds)
 
     # 2
-    ds = set_idomain(ds, remove_nan_layers=remove_nan_layers)
+    if remove_nan_layers:
+        # remove inactive layers
+        ds = remove_inactive_layers(ds)
 
     # 3
+    idomain = get_idomain(ds)
     ds["kh"], ds["kv"] = get_kh_kv(
         ds["kh"],
         ds["kv"],
         anisotropy,
         fill_value_kh=fill_value_kh,
         fill_value_kv=fill_value_kv,
-        idomain=ds["idomain"],
+        idomain=idomain,
     )
     return ds
 
 
-def fill_top_and_bottom(ds):
-    """Remove Nans in botm variable, and change top from 3d to 2d if needed."""
+def fill_top_and_bottom(ds, drop_layer_dim_from_top=True):
```

Comment:
This name is a bit vague considering the change from 3D to 2D top is also done. Maybe we can call this ` to_modflow_tops_and_bottoms()` which calls 2 functions, the first being `fillnan_top_and_bottom()` and the second `make_top_2d()`?

Reply:
I agree in splitting this function. The code to fill nan values in the top and bottom can be optimized using the xarray bfill or ffill methods. I think we should create a separate issue for this

Reply:
I created an issue for this, so we can resolve this conversation.